### PR TITLE
Issue 2519

### PR DESCRIPTION
**Pull request by xhoaluu**
PR for #2519

### CODE REVIEW

**Comment by xhoaluu** [comment]
https://mobicents.ci.cloudbees.com/view/RestComm/job/RestComm-For-Specific-Branch/141/

Test Result (82 failures / -5)
org.restcomm.connect.testsuite.http.LiveCallModificationAnswerDelayTest.holdCall
org.restcomm.connect.testsuite.http.LiveCallModificationAnswerDelayTest.testTerminateDialForkCallWhileRinging_LCM_to_dial_branches
org.restcomm.connect.testsuite.http.LiveCallModificationTest.holdCall
org.restcomm.connect.testsuite.http.LiveCallModificationTest.testTerminateDialForkCallWhileRinging_LCM_to_dial_branches
org.restcomm.connect.testsuite.telephony.CallLifecycleTest.testDialCancelBeforeDialingClientAliceAfterTrying
org.restcomm.connect.testsuite.telephony.CallLifecycleTest.testDialCancelBeforeDialingClientAliceAfterRinging
org.restcomm.connect.testsuite.telephony.CallLifecycleTest.testDialClientAlice
org.restcomm.connect.testsuite.telephony.CallLifecycleTest.testDialClientAliceBehindLB
org.restcomm.connect.testsuite.telephony.CallLifecycleTest.testDialClientAliceWithTimeLimit
org.restcomm.connect.testsuite.telephony.ClientsDialTest.testClientsCallEachOther
org.restcomm.connect.testsuite.telephony.DialForkTest.testDialForkNoAnswerButFromAliceClient
org.restcomm.connect.testsuite.telephony.DialRecordingS3UploadAnswerDelayTest.testDialClientAlice_AliceDisconnects
org.restcomm.connect.testsuite.telephony.DialRecordingS3UploadAnswerDelayTest.testDialClientAlice_BobDisconnects
org.restcomm.connect.testsuite.telephony.DialStatusCallbackTest.testDialForkNoAnswerButHenriqueStatusCallbackOnAll
org.restcomm.connect.testsuite.telephony.push.PushNotificationServerTest.testRcmlCall
org.restcomm.connect.testsuite.provisioning.number.bandwidth.BandwidthIncomingPhoneNumbersEndpointTest.testCancelNumber
org.restcomm.connect.testsuite.provisioning.number.bandwidth.BandwidthIncomingPhoneNumbersEndpointTest.testBuyNumber
org.restcomm.connect.testsuite.provisioning.number.vi.IncomingPhoneNumbersEndpointTest.getIncomingPhoneNumbersList
org.restcomm.connect.testsuite.provisioning.number.vi.IncomingPhoneNumbersEndpointTest.getIncomingPhoneNumbersListUsingPageSize
org.restcomm.connect.testsuite.provisioning.number.vi.IncomingPhoneNumbersEndpointTest.testPurchaseLocalPhoneNumberSuccess
org.restcomm.connect.testsuite.telephony.CallLifecycleAnswerDelayTest.testDialCancelBeforeDialingClientAliceAfterRinging
org.restcomm.connect.testsuite.telephony.DialActionAnswerDelayTest.testDialActionInvalidCall
org.restcomm.connect.testsuite.telephony.DialActionAnswerDelayTest.testDialActionInvalidCallCheckCallStatusCompleted
org.restcomm.connect.testsuite.telephony.ImsClientsDialAnswerDelayCancelTest.testWebRTCClientIncomingCancel
org.restcomm.connect.testsuite.RvdProjectsMigratorWorkspaceMigratedTest.checkIncomingPhoneNumbers
org.restcomm.connect.testsuite.RvdProjectsMigratorWorkspaceMixedTest.checkIncomingPhoneNumbers
org.restcomm.connect.testsuite.RvdProjectsMigratorWorkspaceMixedTest.checkClients
org.restcomm.connect.testsuite.RvdProjectsMigratorWorkspaceOriginalTest.checkIncomingPhoneNumbers
org.restcomm.connect.testsuite.http.AccountsEndpointClosingTest.removeAccountAndReleaseProvidedNumbers
org.restcomm.connect.testsuite.http.AccountsEndpointTest.testCreateAccountWithJapaneseChars
org.restcomm.connect.testsuite.http.CorsRelaxTest.corsHeadersAreReturnedForAccount
org.restcomm.connect.testsuite.http.UsageRecordsTest.getUsageRecordsMonthly
org.restcomm.connect.testsuite.http.UsageRecordsTest.getUsageRecordsYearly
org.restcomm.connect.testsuite.http.UsageRecordsTest.getUsageRecordsAlltime
org.restcomm.connect.testsuite.http.UsageRecordsTest.getUsageRecordsDaily
org.restcomm.connect.testsuite.provisioning.number.voxbone.VoxboneIncomingPhoneNumbersEndpointTest.testPurchasePhoneNumberNoPhoneNumberFound
org.restcomm.connect.testsuite.provisioning.number.voxbone.VoxboneIncomingPhoneNumbersEndpointTest.testPurchaseAndDeletePhoneNumberSuccess
org.restcomm.connect.testsuite.sms.SmsEndpointTest.sendSmsTestToClientExistingInDifferentOrganizations
org.restcomm.connect.testsuite.telephony.ImsClientsDialAnswerDelayTest.testWebRTCClientIncomingADisconnect
org.restcomm.connect.testsuite.telephony.ImsClientsDialAnswerDelayTest.testWebRTCClientIncomingBusy
org.restcomm.connect.testsuite.telephony.ImsClientsDialAnswerDelayTest.testWebRTCClientIncomingFail
org.restcomm.connect.testsuite.telephony.ImsClientsDialAnswerDelayTest.testWebRTCClientIncomingAHold
org.restcomm.connect.testsuite.telephony.ImsClientsDialAnswerDelayTest.testWebRTCClientIncomingBHold
org.restcomm.connect.testsuite.telephony.ImsClientsDialAnswerDelayTest.testWebRTCClientIncomingRequestTimeout
org.restcomm.connect.testsuite.telephony.ImsClientsDialTest.testWebRTCClientOutgoingOtherDomain
org.restcomm.connect.testsuite.telephony.ImsClientsDialTest.testWebRTCClientIncomingADisconnect
org.restcomm.connect.testsuite.telephony.ImsClientsDialTest.testWebRTCClientIncomingBusy
org.restcomm.connect.testsuite.telephony.ImsClientsDialTest.testWebRTCClientIncomingAHold
org.restcomm.connect.testsuite.telephony.ImsClientsDialTest.testWebRTCClientIncomingBHold
org.restcomm.connect.testsuite.telephony.ImsClientsDialTest.testWebRTCClientIncomingRequestTimeout
org.restcomm.connect.testsuite.telephony.NoAuthRegisterClientTest.testGeorgeCallMaria
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceWithRecord2
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialConferenceWithDialActionSms
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialConferenceWithDialActionNoRcml
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialConferenceNoDialAction_SendSms
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialConferenceNoDialAction_SendSms
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialConferenceNoDialAction_NoSms
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceWithRecordAndStatusCallbackForApp
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceWithRecordAndStatusCallbackForAppForThreeCalls
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceWithRecordAndStatusCallbackForAppForThreeCalls
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testRecordWithActionAndStatusCallbackForAppWithDisconnectFromBob
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testRecordWithActionAndStatusCallbackForAppWithBobSendsFinishKey
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testRecordWithActionAndStatusCallbackForAppWithBobSendsFinishKey
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberGeorge
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberGeorgeWithWrongScreeningUrl
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberGeorge_403Forbidden
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberGeorge_404_OnBye
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberGeorgePassInitialCallerId
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberWithPlusSign
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberWithOUTPlusSign
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberNoCountryAccessCode
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberRejectRejectedRcml
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialNumberRejectBusyRcml
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceWithActionAndStatusCallbackForApp
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialTimeOutClientAliceWithActionAndStatusCallbackForApp
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceToBigDID
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwo.testDialClientAliceWithRecord
org.restcomm.connect.testsuite.telephony.TestDialVerbPartTwoAnswerDelay.testDialClientAliceWithRecordAndStatusCallbackForApp
org.restcomm.connect.testsuite.telephony.TestMgcpOperations.testDialHelloPlay
org.restcomm.connect.dao.mybatis.CallDetailRecordsDaoTest.testReadByEndTime
org.restcomm.connect.http.AccountsEndpointMockedTest.requestMissingAuthorizationIsRejected
org.restcomm.connect.testsuite.telephony.ReferTest.org.restcomm.connect.testsuite.telephony.ReferTest

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java [hunk 1]
```diff
@@ -103,13 +104,18 @@ private Address contact(final Gateway gateway, final int expires) throws Servlet
         if (outboundInterface == null)
             outboundInterface = (SipURI) factory.createSipURI(null, address);
         final String user = gateway.getUserName();
+        String contactUser = new String (user);
```
Comment (jaimecasero) [inline restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java]: i think this "new String"  is not strictly neccesary...

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java [hunk 2]
```diff
@@ -20,70 +29,140 @@
...
 public final class ProxyManagerTest {
     private static final String version = Version.getVersion();
 
     @ArquillianResource
     private Deployer deployer;
 
-    private static SipStackTool tool;
-    private SipStack receiver;
-    private SipPhone phone;
+    final String deploymentUrl = "http://127.0.0.1:8080/restcomm/";
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java]: maybe is a good time to refactor thsi test to support parallelization, see other test to check how it goes

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java [hunk 3]
```diff
@@ -20,70 +29,140 @@
...
-        final Request request = register.getRequest();
-        Response ok = factory.createResponse(200, request);
-        SipTransaction transaction = phone.sendReply(register, ok);
-        // Validate the transaction.
-        assertNotNull(transaction);
+        SipURI uri = augustSipStack.getAddressFactory().createSipURI(null, "127.0.0.1:5080");
+        final String userName = "august@127.0.0.1:5092";
+        isAuthorizationHasRightUserName = false;
+        isRegisterRequestReceived = false;
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java]: any reason to do this async to the test thread, why dont ew simple set the Gateway,and then proceed with traffic assertions?

Comment (jaimecasero) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java]: we may need to declare those asserting flags as volatile if several threads are read/writing them..

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java [hunk 4]
```diff
@@ -20,70 +29,140 @@
...
+                    fail(e.getMessage());
+                }catch (InvalidArgumentException e) {
+                    e.printStackTrace();
+                    fail(e.getMessage());
+                }
+
+            }
+        });
+
+        RestcommCallsTool.getInstance().setGateWay(deploymentUrl, adminAccountSid, adminAuthToken, "friendlyName",
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java]: should we asssert something on response, to ensure the response is as expected?

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/http/RestcommCallsTool.java [hunk 5]
```diff
@@ -367,4 +376,42 @@ public JsonArray getCallRecordings(String deploymentUrl, String username, String
...
+        if (proxy != null) {
+            params.add("Proxy", proxy);
+        }
+        if (ttl != null) {
+            params.add("TTL", ttl);
+        }
+
+        String response = null;
+        try {
+            response = webResource.accept(MediaType.APPLICATION_XML).post(String.class, params);
```
Comment (jaimecasero) [inline restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/http/RestcommCallsTool.java]: why are we using XML here, most of the testsuite is using JSON as response format to ease parsing later...

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java [hunk 6]
```diff
@@ -103,13 +104,18 @@ private Address contact(final Gateway gateway, final int expires) throws Servlet
...
+            contactUser = contactUser.substring(0, index);
+        }
         final String host = outboundInterface.getHost();
         int port = outboundInterface.getPort();
         if (port == -1) {
             port = outboundInterface().getPort();
         }
         final StringBuilder buffer = new StringBuilder();
-        buffer.append("sip:").append(user).append("@").append(host).append(":").append(port);
+        buffer.append("sip:").append(contactUser).append("@").append(host).append(":").append(port);
```
Comment (jaimecasero) [inline restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java]: would be great to create meaningful constants fro "@" and ":" is all over the place...

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java (hunk 1 above)
Comment (xhoaluu) [reply restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java]: You are right, to be updated.

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java (hunk 6 above)
Comment (xhoaluu) [reply restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java]: There is a lot of place where is using @ and : .
We need to create a common place to define all of these thing. It's would be great in another story for refactoring code.

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/http/RestcommCallsTool.java (hunk 5 above)
Comment (xhoaluu) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/http/RestcommCallsTool.java]: updated follow comment.

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java (hunk 2 above)
Comment (xhoaluu) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java]: updated to parallel test

File: restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java (hunk 6 above)
Comment (jaimecasero) [reply restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/proxy/ProxyManager.java]: agree

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java (hunk 3 above)
Comment (xhoaluu) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java]: As the time, copied pasted without evaluating. 
I removed the parallel code and also removed the check variables, put direct assert at the point as we don't use thread any more.

File: restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java (hunk 4 above)
Comment (xhoaluu) [reply restcomm/restcomm.testsuite/src/test/java/org/restcomm/connect/testsuite/telephony/proxy/ProxyManagerTest.java]: As response is from simulator, then nothing to check there.
I changed the testcase name, then it's cleared about the target for the testcase.
I just need to check if username has host part, REGISTER can be sent out and the username in authentication header in REGISTER request is what we want.

**Pull request closed without merging**